### PR TITLE
Seed IBLT hash functions

### PR DESCRIPTION
**Pull request by bissias**
Previously, in the construction of a CGrapheneSet, the ith hash function operating on input x always evaluated to the same value across all instantiations. This is undesirable because IBLT decode failures, which are pseudo-random, have the potential to occur for multiple peers when reconciling the same block.

The present PR adds the ability to pass the CIblt object an arbitrary 32-bit salt value. This salt is used to seed the IBLT hash functions. For entropy, we use the random nonce created by the Graphene block sender when seeding the SipHash used for Graphene cheap hashes.

### CODE REVIEW

**Comment by sickpig** [comment]
@bissias PR needs a rebase.

File: src/blockrelay/graphene_set.h
Comment (ptschip) [inline src/blockrelay/graphene_set.h]: this mutable seems to have returned...did you correctly rebase this PR on top of the previous one ?

File: src/blockrelay/graphene.cpp
Comment (ptschip) [inline src/blockrelay/graphene.cpp]: since GetShortID isn't static and GetShortID is also defined in compactblocks.cpp/h.  I think it would be better to either make the distinction here and call this GetGrShortID() or perhaps better , combine the two ...create a common GetShortID() and put it in blockrelay_common and pass the byte size you want since compactblocks uses 6 and graphene uses 8.

File: src/blockrelay/graphene.cpp
Comment (ptschip) [inline src/blockrelay/graphene.cpp:55]: this is a little untitutive to me..there is a version , then a graphenesetversion then further down there is grapheneVersion?  are they all different or is version also == grapheneVersion?  if version means grapheneVersion then maybe call it that...would make the code more readable IMHO.

File: src/blockrelay/graphene_set.h
Comment (bissias) [reply src/blockrelay/graphene_set.h]: Shoot, that must be the case; will fix. The rebase has been absolutely brutal since this is based off of #1551 and they change so many of the same things (~10 rounds of merge errors with 10s of chunks in multiple files each round). I might wait until #1551 is merged so I only have to rebase against its changes one more time.

File: src/blockrelay/graphene.cpp
Comment (bissias) [reply src/blockrelay/graphene.cpp]: I'm not strongly opposed to putting `GetShortID` in blockrelay_common, but in addition to differing byte count, graphene also need to do a version check. So the only line that remains the same is the last one. For that reason, I lean toward only renaming it to `GetGrShortID` as you suggested.

File: src/blockrelay/graphene.cpp
Comment (bissias) [reply src/blockrelay/graphene.cpp:55]: I agree that it's cumbersome, but let me explain the reasoning a little; then perhaps you can suggest a better solution? There are now three objects related to graphene that have a (separate) version number: `CGrapheneBlock`, `CGrapheneSet`, and `CIblt`. I thought about just using the same version number for each, but it ended up not making sense to me because then you would need to, for example, bump the version of `CIblt` every time you make a change to `CGrapheneBlock`. So that's my reasoning for having two different variables: `grapheneVersion` and `grapheneSetVersion`, in general they are different. Regarding the generic name `version`, each one of those objects uses `version` as a member variable where it is unambiguous.  I use a more descriptive variable name in cases where the class associated with the version is not necessarily clear. Does that makes sense?

**Comment by ptschip** [comment]
ok i see.

File: src/blockrelay/graphene.cpp
Comment (gandrewstone) [reply src/blockrelay/graphene.cpp:55]: on a related note, if grapheneSetVersion is changing from a prior recent PR to this one, we can combine that into the same version number, since we did not make a release or merge to our release branch.

File: src/blockrelay/graphene_set.cpp
Comment (gandrewstone) [inline src/blockrelay/graphene_set.cpp]: is MaxNHash guaranteed to be != 0?

File: src/iblt.h
Comment (gandrewstone) [inline src/iblt.h:153]: add a doc comment (//! or /** here) describing what salt does and why.  A subset of what you put in this PR description is what I'm looking for.

**Comment by sickpig** [comment]
@bissias need a rebase.

File: src/blockrelay/graphene_set.cpp
Comment (bissias) [reply src/blockrelay/graphene_set.cpp]: The way it's currently setup in iblt.cpp, the smallest it can be is 4. But I redefined ibltSalt so that the divisor in the modulus is no less than 1, just to be safe.

File: src/iblt.h
Comment (bissias) [reply src/iblt.h:153]: Done.

File: src/blockrelay/graphene.cpp
Comment (bissias) [reply src/blockrelay/graphene.cpp:55]: I would rather not do that only because version number checks are performed in a few different places, leaving room to introduce an error if changed. Also, downstream patches based on this one bump the version number assuming the current value.

Comment (gandrewstone) [reply src/blockrelay/graphene.cpp:55]: no problem

File: src/iblt.cpp
Comment (gandrewstone) [inline src/iblt.cpp]: use std::numeric_limits<uint32_t>::max()

File: src/iblt.cpp
Comment (gandrewstone) [inline src/iblt.cpp]: biggest number is salt * n_hash + n_hash - 1.  Replacing salt with the check above -> (VALS_32 / n_hash * n_hash) + n_hash -1, or VALS_32 + n_hash - 1.  So your check is too lenient by n_hash.

File: src/blockrelay/graphene_set.h
Comment (gandrewstone) [inline src/blockrelay/graphene_set.h:162]: can I give you a salt that much later triggers the iblt.h line 142 exception (salt too big based on n_hashes)?

File: src/iblt.cpp
Comment (gandrewstone) [inline src/iblt.cpp:161]: for consistency add & KEYCHECK_MASK here or remove it from the above (since its a nop because the return value is defined as a uint32_t).

File: src/iblt.h
Comment (gandrewstone) [inline src/iblt.h]: remove (replace uses with the standard constant as described in another comment)

File: src/iblt.h
Comment (gandrewstone) [inline src/iblt.h]: move this test lower, since it is using n_hash which has not been read yet.

File: src/blockrelay/graphene_set.cpp
Comment (gandrewstone) [inline src/blockrelay/graphene_set.cpp]: This is pedantic because MaxNHash is so small.  But if MaxNHash was > 65k, and certain ibltEntropy numbers were chosen, then ibltSalt could be > 65k, resulting in a salt*n_hashes > max int.  Shouldn't this eqn be:

ibltEntropy % (std::max<uint32_t>/std::max(1, (int)CIblt::MaxNHash());

A practical effect of using your eqn is that ibltSalt can really only vary from 0 to 8.  This isn't a lot of entropy for the effort.  I think we could have more.  But at the same time I don't like my eqn because there's no safety factor.  It would be better to ensure that throwing when salt*n_hashes > 0xffffffff will never happen in practice, except for nodes that give garbage values.

File: src/iblt.cpp
Comment (gandrewstone) [inline src/iblt.cpp]: do we really need this, or can mapHashIdxSeeds wrap around?

**Review by gandrewstone — CHANGES_REQUESTED** [review]

File: src/blockrelay/graphene_set.cpp
Comment (bissias) [reply src/blockrelay/graphene_set.cpp]: Your comment made me take a second look at the logic in iblt.cpp that constructs hash seeds from the salt. When I originally wrote that code, I had a reason to ensure that different salts would produce disjoint sets of seeds. But that is no longer important, so I think this much simpler logic for seed `i` will suffice: `i + salt % (0xffffffff - n_hash)`, which ensures distinct seeds for each hash function. In light of that change, it is no longer necessary to perform any range checks on the salt. So I believe the line in question here becomes simply `ibltSalt = ibltEntropy`.

File: src/iblt.cpp
Comment (bissias) [reply src/iblt.cpp]: Removed in 4eee6a1.

File: src/iblt.h
Comment (bissias) [reply src/iblt.h]: Removed entirely in 4eee6a1.

File: src/iblt.h
Comment (bissias) [reply src/iblt.h]: Done.

File: src/iblt.cpp
Comment (bissias) [reply src/iblt.cpp]: Good point, but this line has been changed in 4eee6a1.

File: src/iblt.cpp
Comment (bissias) [reply src/iblt.cpp]: Removed entirely in 4eee6a1.

File: src/blockrelay/graphene_set.h
Comment (bissias) [reply src/blockrelay/graphene_set.h:162]: With 4eee6a1 it should not be possible because any 32-bit salt value is valid for the IBLT.

File: src/iblt.cpp
Comment (bissias) [reply src/iblt.cpp:161]: The use of KEYCHECK_MASK here was extraneous, so I removed it.

**Review by gandrewstone — APPROVED** [review]

**Pull request merged**
